File: internal/models/subscriptions.go
// agency-gateway/internal/models/subscriptions.go
package models

import "fmt"

// MatchClassification constants for subscription matching.
const (
	MatchClassificationDirect        = "direct"
	MatchClassificationInterestMatch = "interest_match"
	MatchClassificationAmbient       = "ambient"
)

// ExpertiseTier constants for expertise declaration tiers.
const (
	ExpertiseTierBase     = "base"
	ExpertiseTierStanding = "standing"
	ExpertiseTierLearned  = "learned"
	ExpertiseTierTask     = "task"
)

// ExpertiseDeclaration declares an agent's expertise in a domain.
type ExpertiseDeclaration struct {
	Tier        string   `yaml:"tier" json:"tier"`
	Description string   `yaml:"description" json:"description"`
	Keywords    []string `yaml:"keywords" json:"keywords"`
	Persistent  bool     `yaml:"persistent" json:"persistent"`
}

// Validate checks keyword count and filters short keywords (< 3 chars).
func (e *ExpertiseDeclaration) Validate() error {
	if len(e.Keywords) > 30 {
		return fmt.Errorf("expertise declarations support at most 30 keywords")
	}
	filtered := e.Keywords[:0]
	for _, kw := range e.Keywords {
		if len(kw) >= 3 {
			filtered = append(filtered, kw)
		}
	}
	e.Keywords = filtered
	return nil
}

// InterestDeclaration declares an agent's interest in a topic for a task.
type InterestDeclaration struct {
	TaskID          string              `yaml:"task_id" json:"task_id"`
	Description     string              `yaml:"description" json:"description"`
	Keywords        []string            `yaml:"keywords" json:"keywords"`
	KnowledgeFilter map[string][]string `yaml:"knowledge_filter" json:"knowledge_filter"`
}

// Validate checks keyword count, filters short keywords, and checks knowledge_filter total.
func (d *InterestDeclaration) Validate() error {
	if len(d.Keywords) > 20 {
		return fmt.Errorf("interest declarations support at most 20 keywords")
	}
	filtered := d.Keywords[:0]
	for _, kw := range d.Keywords {
		if len(kw) >= 3 {
			filtered = append(filtered, kw)
		}
	}
	d.Keywords = filtered

	total := 0
	for _, entries := range d.KnowledgeFilter {
		total += len(entries)
	}
	if total > 10 {
		return fmt.Errorf("knowledge filter supports at most 10 entries")
	}
	return nil
}

// WSEvent is a WebSocket event envelope sent to subscribed clients.
type WSEvent struct {
	V               int                    `yaml:"v" json:"v"`
	Type            string                 `yaml:"type" json:"type"`
	Channel         *string                `yaml:"channel" json:"channel,omitempty"`
	Match           *string                `yaml:"match" json:"match,omitempty"`
	MatchedKeywords []string               `yaml:"matched_keywords" json:"matched_keywords,omitempty"`
	Message         map[string]interface{} `yaml:"message" json:"message,omitempty"`
	Task            map[string]interface{} `yaml:"task" json:"task,omitempty"`
	Event           *string                `yaml:"event" json:"event,omitempty"`
	Data            map[string]interface{} `yaml:"data" json:"data,omitempty"`
}

// InterruptionRule defines how a comms match classification is handled.
type InterruptionRule struct {
	Match  string   `yaml:"match" json:"match"`
	Flags  []string `yaml:"flags" json:"flags"`
	Action string   `yaml:"action" json:"action"`
}

// CommsPolicy defines how incoming messages interrupt or queue for an agent.
type CommsPolicy struct {
	Rules                       []InterruptionRule `yaml:"rules" json:"rules"`
	MaxInterruptsPerTask        int                `yaml:"max_interrupts_per_task" json:"max_interrupts_per_task"`
	CooldownSeconds             int                `yaml:"cooldown_seconds" json:"cooldown_seconds"`
	IdleAction                  string             `yaml:"idle_action" json:"idle_action"`
	CircuitBreakerMinActionRate float64            `yaml:"circuit_breaker_min_action_rate" json:"circuit_breaker_min_action_rate"`
	CircuitBreakerWindowSize    int                `yaml:"circuit_breaker_window_size" json:"circuit_breaker_window_size"`
}

// DefaultCommsPolicy returns a CommsPolicy with the standard four-rule default.
func DefaultCommsPolicy() CommsPolicy {
	return CommsPolicy{
		Rules: []InterruptionRule{
			{Match: MatchClassificationDirect, Flags: []string{"urgent", "blocker"}, Action: "interrupt"},
			{Match: MatchClassificationDirect, Flags: []string{}, Action: "notify_at_pause"},
			{Match: MatchClassificationInterestMatch, Flags: []string{}, Action: "notify_at_pause"},
			{Match: MatchClassificationAmbient, Flags: []string{}, Action: "queue"},
		},
		MaxInterruptsPerTask:        3,
		CooldownSeconds:             60,
		IdleAction:                  "queue",
		CircuitBreakerMinActionRate: 0.2,
		CircuitBreakerWindowSize:    20,
	}
}

File: internal/images/contexts/_shared/models/subscriptions.py
"""Models for real-time comms subscriptions, matching, and interruption policy.

Supports WebSocket push events, interest declarations with keyword/knowledge
filtering, and operator-defined interruption policies.
"""

from enum import Enum
from typing import Any

from pydantic import BaseModel, Field, field_validator


class MatchClassification(str, Enum):
    DIRECT = "direct"
    INTEREST_MATCH = "interest_match"
    AMBIENT = "ambient"


class ExpertiseTier(str, Enum):
    BASE = "base"
    STANDING = "standing"
    LEARNED = "learned"
    TASK = "task"


class ExpertiseDeclaration(BaseModel):
    """A single tier of an agent's expertise profile.

    Four tiers:
    - base: operator-defined in agent.yaml (read-only to agent, tenet 5)
    - standing: operator-assigned via task delivery ("help anyone who asks about X")
    - learned: accumulated from past work (agent-managed)
    - task: current task focus (temporary, cleared on task end)
    """
    tier: ExpertiseTier
    description: str = ""
    keywords: list[str] = Field(default_factory=list)
    persistent: bool = False

    @field_validator("keywords")
    @classmethod
    def validate_keywords(cls, v: list[str]) -> list[str]:
        if len(v) > 30:
            raise ValueError("Expertise declarations support at most 30 keywords")
        return [kw for kw in v if len(kw) >= 3]


# Keep InterestDeclaration as alias for backward compatibility during migration
class InterestDeclaration(BaseModel):
    task_id: str
    description: str = ""
    keywords: list[str] = Field(default_factory=list)
    knowledge_filter: dict[str, list[str]] = Field(default_factory=dict)

    @field_validator("keywords")
    @classmethod
    def validate_keywords(cls, v: list[str]) -> list[str]:
        if len(v) > 20:
            raise ValueError("Interest declarations support at most 20 keywords")
        return [kw for kw in v if len(kw) >= 3]

    @field_validator("knowledge_filter")
    @classmethod
    def validate_knowledge_filter(cls, v: dict[str, list[str]]) -> dict[str, list[str]]:
        total = sum(len(entries) for entries in v.values())
        if total > 10:
            raise ValueError(
                "Knowledge filter supports at most 10 entries (kinds + topics combined)"
            )
        return v


class WSEvent(BaseModel):
    v: int = 1
    type: str
    channel: str | None = None
    match: str | None = None
    matched_keywords: list[str] | None = None
    message: dict[str, Any] | None = None
    task: dict[str, Any] | None = None
    event: str | None = None
    data: dict[str, Any] | None = None


class InterruptionRule(BaseModel):
    match: str
    flags: list[str] = Field(default_factory=list)
    action: str = "queue"


class CommsPolicy(BaseModel):
    rules: list[InterruptionRule] = Field(default_factory=lambda: [
        InterruptionRule(match="direct", flags=["urgent", "blocker"], action="interrupt"),
        InterruptionRule(match="direct", action="notify_at_pause"),
        InterruptionRule(match="interest_match", action="notify_at_pause"),
        InterruptionRule(match="ambient", action="queue"),
    ])
    max_interrupts_per_task: int = 3
    cooldown_seconds: int = 60
    idle_action: str = "queue"
    circuit_breaker_min_action_rate: float = 0.2
    circuit_breaker_window_size: int = 20
